prov/rxm/src/rxm_conn.c

```diff
@@ -431,14 +431,19 @@ static int rxm_cmap_handle_connect(struct rxm_cmap *cmap, fi_addr_t fi_addr,
 	case RXM_CMAP_CONNECTED:
 		return FI_SUCCESS;
 	case RXM_CMAP_IDLE:
-		ret = rxm_conn_connect(cmap->ep, handle,
-				       ofi_av_get_addr(cmap->av, fi_addr),
-				       cmap->av->addrlen);
-		if (ret) {
-			rxm_cmap_del_handle(handle);
-			return ret;
+		if (cmap->conn_quota > 0) {
+			ret = rxm_conn_connect(cmap->ep, handle,
+					       ofi_av_get_addr(cmap->av, fi_addr),
+					       cmap->av->addrlen);
+			if (ret) {
+				rxm_cmap_del_handle(handle);
+				return ret;
+			}
+			handle->state = RXM_CMAP_CONNREQ_SENT;
+		} else if (slist_entry_empty(&handle->wait_conn_quota_entry)) {
+			slist_insert_tail(&handle->wait_conn_quota_entry,
+					  &cmap->conn_wait_quota_list);
```

Comment:
Something looks off here.  What happens if conn_quota == 0, but the list isn't empty?

Reply:
I finally figured this out after reading further down in the code.  I even tried to come up with better name than slist_entry_empty(), but nothing was sticking.  I think the check that's being done is along the lines of slist_entry_inlist()?  But I'm not sure if that really works.  The check is whether the pointer is null, which would also be the case if the entry is at the tail of the list.  So, maybe the logic is off here.

Reply:
May be use a dlist instead?